js/pro/gemini.js

```diff
@@ -0,0 +1,663 @@
+            this.ohlcvs[symbol] = {};
+        }
+        let stored = this.safeValue (this.ohlcvs[symbol], timeframe);
+        if (stored === undefined) {
+            const limit = this.safeInteger (this.options, 'OHLCVLimit', 1000);
+            stored = new ArrayCacheByTimestamp (limit);
+            this.ohlcvs[symbol][timeframe] = stored;
+        }
+        const changesLength = changes.length;
+        // reverse order of array to store candles in ascending order
+        for (let i = changesLength - 1; i >= 0; i--) {
+            const parsed = this.parseOHLCV (changes[i], market);
+            stored.append (parsed);
+        }
+        const messageHash = 'ohlcv:' + symbol + ':' + timeframeId;
+        client.resolve (stored, messageHash);
+        return message;
+    }
+
+    async watchOrderBook (symbol, limit = undefined, params = {}) {
+        /**
+         * @method
+         * @name gemini#watchOrderBook
+         * @description watches information on open orders with bid (buy) and ask (sell) prices, volumes and other data
+         * @see https://docs.gemini.com/websocket-api/#market-data-version-2
+         * @param {string} symbol unified symbol of the market to fetch the order book for
+         * @param {int|undefined} limit the maximum amount of order book entries to return
+         * @param {object} params extra parameters specific to the gemini api endpoint
+         * @returns {object} A dictionary of [order book structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-book-structure} indexed by market symbols
+         */
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        const messageHash = 'orderbook:' + market['symbol'];
+        const marketId = market['id'];
+        const request = {
+            'type': 'subscribe',
+            'subscriptions': [
+                {
+                    'name': 'l2',
+                    'symbols': [
+                        marketId.toUpperCase (),
+                    ],
+                },
+            ],
+        };
+        const subscribeHash = 'l2:' + market['symbol'];
+        const url = this.urls['api']['ws'] + '/v2/marketdata';
+        const orderbook = await this.watch (url, messageHash, request, subscribeHash);
+        return orderbook.limit (limit);
+    }
+
+    handleOrderBook (client, message) {
+        const changes = this.safeValue (message, 'changes', []);
+        const marketId = this.safeStringLower (message, 'symbol');
+        const market = this.safeMarket (marketId);
+        const symbol = market['symbol'];
+        const messageHash = 'orderbook:' + symbol;
+        let orderbook = this.safeValue (this.orderbooks, symbol);
+        if (orderbook === undefined) {
+            orderbook = this.orderBook ();
+        }
+        for (let i = 0; i < changes.length; i++) {
+            const delta = changes[i];
+            const price = this.safeFloat (delta, 1);
```

Comment:
@pcriadoperez btw we should avoid using `safeFloat`because it does not take into consideration the precision mode, `safeNumber` is preferred